### PR TITLE
[3.3.5] Implement Pet combo point system

### PR DESCRIPTION
**Pull request by Treeston**
* Implement Pet combo point system
* Implement a hack to add combo points and make Wolverine Bite useful (fixes #752).
* Fix an issue where players could use client hacks to use combo points on a different target from the one they're applied to.

### CODE REVIEW

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -2106,6 +2106,12 @@ void Unit::AttackerStateUpdate(Unit* victim, WeaponAttackType attType, bool extr
         DamageInfo dmgInfo(damageInfo);
         ProcSkillsAndAuras(damageInfo.target, damageInfo.procAttacker, damageInfo.procVictim, PROC_SPELL_TYPE_NONE, PROC_SPELL_PHASE_NONE, dmgInfo.GetHitMask(), nullptr, &dmgInfo, nullptr);
 
+        if (dmgInfo.GetHitMask() & PROC_HIT_CRITICAL) // @todo is there a better place to put this hack? there doesn't seem to be any aura applied on talent
+            if (Pet* pet = ToPet())
+                if (pet->HasSpell(53508)) // Wolverine Bite
+                    pet->AddComboPoints(damageInfo.target, 1);
```
Comment (Treeston) [inline src/server/game/Entities/Unit/Unit.cpp]: This segment is obviously a hack. However, I didn't find any auras applied by the talent - in fact, I didn't even find any global pet auras that proc on white crit - despite my best efforts.

@Shauren any ideas on how to best handle this? Leaving it here feels wrong.

Comment (ccrs) [inline src/server/game/Entities/Unit/Unit.cpp]: script hook and playerscript?

Comment (ariel-) [inline src/server/game/Entities/Unit/Unit.cpp]: Unit::ProcSkillsAndReactives seems a better place for this

**Comment by ariel-** [comment]
What about making combo point related methods virtual? so only one call instead of branching

**Comment by Treeston** [comment]
They're only implemented by `Pet` and `Player`. Not sure what's cheaper here - @Shauren, insights?

File: src/server/game/Entities/Unit/Unit.cpp (hunk 1 above)
Comment (ariel-) [inline src/server/game/Entities/Unit/Unit.cpp]: Also, check for special attribute in Wolverine Bite (IIRC had one)

Comment (Treeston) [inline src/server/game/Entities/Unit/Unit.cpp]: It has `SPELL_ATTR4_TRIGGER_ACTIVATE` and `SPELL_ATTR1_UNK27`, no idea if either of those is interesting.

Comment (ariel-) [inline src/server/game/Entities/Unit/Unit.cpp]: SPELL_ATTR4_TRIGGER_ACTIVATE check that one :)

Comment (Treeston) [inline src/server/game/Entities/Unit/Unit.cpp]: `TRIGGER_ACTIVATE` is a client side flag that interacts with other flags (in this case `REQ_COMBO_POINTS1`) to make the icon gray out if not available. No server handling for these spells.

NB: Most other spells with this attribute have proc info (f.ex. Art of War).

**Comment by Shauren** [comment]
In this case? Whatever produces cleaner code

File: src/server/game/Entities/Unit/Unit.cpp (hunk 1 above)
Comment (Shauren) [inline src/server/game/Entities/Unit/Unit.cpp]: Lets see, a spell using a single combo point to activate, does not scale from combo points...
Yep, thats a reactive, just like warrior Overpower - handle them together in Unit::ProcSkillsAndReactives

Comment (ariel-) [inline src/server/game/Entities/Unit/Unit.cpp]: I recall researching for the WB issue, I think SMSG_PET_UPDATE_COMBO_POINTS was sent even when the pet didn't have the spell, so a mere AddComboPoints on every crit should do in that case

**Comment by Treeston** [comment]
OK, I'll extend this to cover a full CP system rewrite then.

**Comment by Treeston** [comment]
PR superseded by #19914.

**Pull request closed without merging**